routers/web/repo/actions/badge.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package actions
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net/http"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	actions_model "code.gitea.io/gitea/models/actions"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/util"
+
+	"golang.org/x/image/font"
+	"golang.org/x/image/font/basicfont"
+)
+
+type Badge struct {
+	Label         string
+	Message       string
+	Color         string
+	Width         int
+	LabelWidth    int
+	MessageWidth  int
+	LabelX        int
+	MessageX      int
+	LabelLength   int
+	MessageLength int
+	MessageColor  string
+	MessageShadow string
+	FontSize      int
+}
```

Comment:
```suggestion
type BadgeSize struct {
	Width  int
	Height int
}

type BadgeComponentLocation struct {
	X int
	Y int
}

type BadgeBase struct {
	Color string
	Size  BadgeSize
}

type BadgeComponent struct {
	BadgeBase
	Text       string
	TextLength int
	Location   BadgeComponentLocation
}

type BadgeLabel struct {
	BadgeComponent
}

type BadgeMessage struct {
	BadgeComponent
	Shadow string
}

type Badge struct {
	BadgeBase
	FontSize int

	// Components
	Label   BadgeLabel
	Message BadgeMessage
}
```

Maybe this is more readable.